Refactor v2 spi

bradleyharden opened the pull request.

jbeaurivage commented on hal/src/sercom/v2/spi.rs:
Is it really `unsafe`? In a previous discussion we agreed that `unsafe` should, as much as possible, be used only for things that may incur memory unsafety/UB. I guess it could be argued that this is "memory unsafe" as the data received is not what was expected. I'm not sure.

jbeaurivage commented on hal/src/sercom/v2/spi/impl_ehal_thumbv6m.rs:
```diff
@@ -0,0 +1,475 @@
+//! `CharSize`. For example, an [`Spi`] struct with a `NineBit` `CharSize` would
+//! implement `spi::FullDuplex<u16>`.
+//!
+//! Note that embedded HAL does not offer a way to transmit slices in a
+//! non-blocking fashion, but this can be done using interrupts and the
+//! [`spi_future`](super::super::spi_future) module.
```
Maybe mention that DMA may also be used?

jbeaurivage commented on hal/src/sercom/v2/spi/impl_ehal_thumbv7em.rs:
```diff
@@ -0,0 +1,750 @@
+//! Because a `DynLength` is not guaranteed to be an `AtomicSize`, the
+//! corresponding `Spi` structs only implement the `blocking` traits as well.
+//!
+//! For a non-blocking alternative that can be used to transfer arbitrary-length
+//! slices, see the [`spi_future`](super::super::spi_future) module.
```
Again, might be nice to mention DMA. (I would advise against feature-gating this behind the `dma` Cargo feature. We want the users to be able to find out that it's a possibility without explicitly having to turn it on beforehand).

jbeaurivage commented on hal/src/sercom/v2/spi/impl_ehal_thumbv7em.rs:
Is it normal that there isn't an `AtomicSize` trait bound here?

jbeaurivage replied:
Ah, or `PrimInt + AsPrimitive<u32>` takes care of it?

jbeaurivage commented on hal/src/sercom/v2/spi/pads_thumbv6m.rs:
```diff
@@ -0,0 +1,517 @@
+    (@dopo, $DI:ident, NoneT, Pad1, $SS:ident, $DIPO:literal) => { impl_dipodopo!(@filter, $DI, NoneT, Pad1, $SS, $DIPO, 0); };
+    (@dopo, $DI:ident, NoneT, Pad3, $SS:ident, $DIPO:literal) => { impl_dipodopo!(@filter, $DI, NoneT, Pad3, $SS, $DIPO, 1); };
+    (@dopo, $DI:ident, Pad0,  Pad1, $SS:ident, $DIPO:literal) => { impl_dipodopo!(@filter, $DI, Pad0,  Pad1, $SS, $DIPO, 0); };
+    (@dopo, $DI:ident, Pad2,  Pad3, $SS:ident, $DIPO:literal) => { impl_dipodopo!(@filter, $DI, Pad2,  Pad3, $SS, $DIPO, 1); };
+    (@dopo, $DI:ident, Pad3,  Pad1, $SS:ident, $DIPO:literal) => { impl_dipodopo!(@filter, $DI, Pad3,  Pad1, $SS, $DIPO, 2); };
+    (@dopo, $DI:ident, Pad0,  Pad3, $SS:ident, $DIPO:literal) => { impl_dipodopo!(@filter, $DI, Pad0,  Pad3, $SS, $DIPO, 3); };
+
+    // If DO is not valid, fall through to this pattern.
+    (@dopo, $DI:ident, $DO:ident, $CK:ident, $SS:ident, $DIPO:literal) => { };
+
+    // Filter any remaining permutations that conflict.
+    (@filter, NoneT, NoneT, $CK:ident, $SS:ident, $DIPO:literal, $DOPO:literal) => { };
+    (@filter, Pad0, Pad0, $CK:ident, $SS:ident, $DIPO:literal, $DOPO:literal) => { };
+    (@filter, Pad1, $DO:ident, Pad1, $SS:ident, $DIPO:literal, $DOPO:literal) => { };
+    (@filter, Pad1, $DO:ident, $CK:ident, Pad1, $DIPO:literal, $DOPO:literal) => { };
+    (@filter, Pad2, Pad2, $CK:ident, $SS:ident, $DIPO:literal, $DOPO:literal) => { };
+    (@filter, Pad2, $DO:ident, $CK:ident, Pad2, $DIPO:literal, $DOPO:literal) => { };
+    (@filter, Pad3, Pad3, $CK:ident, $SS:ident, $DIPO:literal, $DOPO:literal) => { };
+    (@filter, Pad3, $DO:ident, Pad3, $SS:ident, $DIPO:literal, $DOPO:literal) => { };
+    (@filter, $DI:ident, Pad2, $CK:ident, Pad2, $DIPO:literal, $DOPO:literal) => { };
+
+    // If there are no conflicts, fall through to this pattern
+    (@filter, $DI:ident, $DO:ident, $CK:ident, $SS:ident, $DIPO:literal, $DOPO:literal) => { impl_dipodopo!(@revise, $DI, $DO, $CK, $SS, $DIPO, $DOPO); };
+
+    // Revise DIPO and DOPO conflicts created by NoneT
+    (@revise, NoneT, Pad0, Pad1, $SS:ident, $DIPO:literal, $DOPO:literal) => { impl_dipodopo!(@implement, NoneT, Pad0, Pad1, $SS, 3, $DOPO); };
+    (@revise, NoneT, Pad0, Pad3, $SS:ident, $DIPO:literal, $DOPO:literal) => { impl_dipodopo!(@implement, NoneT, Pad0, Pad3, $SS, 2, $DOPO); };
+    (@revise, Pad0, NoneT, Pad1, $SS:ident, $DIPO:literal, $DOPO:literal) => { impl_dipodopo!(@implement, Pad0, NoneT, Pad1, $SS, $DIPO, 2); };
+    (@revise, Pad2, NoneT, Pad3, $SS:ident, $DIPO:literal, $DOPO:literal) => { impl_dipodopo!(@implement, Pad2, NoneT, Pad3, $SS, $DIPO, 3); };
+
+    // If there are no revisions, fall through to this pattern
+    (@revise, $DI:ident, $DO:ident, $CK:ident, $SS:ident, $DIPO:literal, $DOPO:literal) => { impl_dipodopo!(@implement, $DI, $DO, $CK, $SS, $DIPO, $DOPO); };
+
+    // Implement DipoDopo
+    (@implement, $DI:ident, $DO:ident, $CK:ident, $SS:ident, $DIPO:literal, $DOPO:literal) => {
+        impl DipoDopo for ($DI, $DO, $CK, $SS) {
+            const DIPO_DOPO: (u8, u8) = ($DIPO, $DOPO);
+        }
+    };
+}
```
And they say C macros are "fun"!

jbeaurivage commented on hal/src/sercom/v2/spi/pads_thumbv7em.rs:
```diff
@@ -0,0 +1,385 @@
+// Dipo
+//=============================================================================
+
+/// Map an [`OptionalPadNum`] to its corresponding `DIPO` value
+pub trait Dipo: OptionalPadNum {
+    const DIPO: Option<u8>;
+}
+
+impl Dipo for NoneT {
+    const DIPO: Option<u8> = None;
+}
+
+impl Dipo for Pad0 {
+    const DIPO: Option<u8> = Some(0);
+}
+impl Dipo for Pad1 {
+    const DIPO: Option<u8> = Some(1);
+}
+impl Dipo for Pad2 {
+    const DIPO: Option<u8> = Some(2);
+}
+impl Dipo for Pad3 {
+    const DIPO: Option<u8> = Some(3);
+}
```
This is new. Is there a downside to just setting `DIPO/DOPO` to a nonsensical value instead? (asking for `uart-v2`).

bradleyharden commented:
@jacobrosenthal, I refactored this to get rid of its dependence on the pygamer port. I also fixed DMA in light of the UART merge. I still need to go through your comments.

bradleyharden commented:
@jbeaurivage, see the latest commit to address your comments

bradleyharden commented:
I had to make some changes to `metro_m4`'s Cargo.toml and run `cargo update` to get it to compile. Not sure what happened there.

bradleyharden commented:
Whoops, I already know this is going to fail. I forgot to update the `metro_m4` for the new `Capability` feature.
